### DISCUSSION TITLE
How to use simple-websocket with this package

### DISCUSSION THREAD

**Discussion by jhoogstraat** [miguelgrinberg/Flask-SocketIO#1732] [Q&A]
How would one use the mentioned websocket server with this package? The documentation does not go into detail. No information is given when booting the server either. It just does not advertise a websocket upgrade.

I need to use this because of the threading async_mode.

**Comment by miguelgrinberg** [comment]
You just have to install `simple-websocket` in your virtual environment. There is nothing else to do.

**Comment by jhoogstraat** [comment]
Then something else must be wrong.

The server does not advertise a websocket upgrade (aka `upgrades: []`), which it should.
I am unsure on how to debug this further. Do you have a suggestion?

**Reply by miguelgrinberg** [reply]
Do you have the code and logs? If you don't know how to enable logs, see the troubleshooting section of the docs.

**Reply by jhoogstraat** [reply]
I currently don't have access to the hardware, but the logs didn't show anything related to this issue (or any error at all) using `logger=True, engineio_logger=True`

I tried starting the server:
- Using the `.run()` method
- Using gunicorn

With both I tried:
- connecting from the js socketio client library
- a dart socketio client package

but neither were promoted to websocket from simple polling.

**Reply by miguelgrinberg** [reply]
There should have been something. The server should have indicated that upgrade is enabled, and the client must have requested the upgrade. If the server advertised the upgrade, but the client didn't take the offer, then the problem is with the client.

**Reply by jhoogstraat** [reply]
I was finally able to get a log.
As you can see there are no indicators why websocket is not available.

That is using `async_mode='threading'`
and
`gunicorn3 --threads 8 -w 1 -b 0.0.0.0:5000 'app:app'`:

```
[2021-11-24 09:17:15 +0000] [15658] [INFO] Starting gunicorn 19.9.0
[2021-11-24 09:17:15 +0000] [15658] [INFO] Listening at: http://0.0.0.0:5000 (15658)
[2021-11-24 09:17:15 +0000] [15658] [INFO] Using worker: threads
[2021-11-24 09:17:16 +0000] [15672] [INFO] Booting worker with pid: 15672
Server initialized for threading.
4lVxRsjzFJJ-qGl6AAAA: Sending packet OPEN data {'sid': '4lVxRsjzFJJ-qGl6AAAA', 'upgrades': [], 'pingTimeout': 5000, 'pingInterval': 25000}
4lVxRsjzFJJ-qGl6AAAA: Received packet MESSAGE data 0
4lVxRsjzFJJ-qGl6AAAA: Sending packet MESSAGE data 0{"sid":"DKJiDUnfJzWsfaWKAAAB"}
```

`pip list`:
```

Flask                            1.1.2
Flask-SocketIO                   5.1.1
gevent                           20.9.0
greenlet                         0.4.17
gunicorn                         19.9.0
Werkzeug                         2.0.2
simple-websocket                 0.4.0
```

**Reply by miguelgrinberg** [reply]
Are you sure the client isn't directly connecting over WebSocket? That's what it looks like to me. The `upgrades: []` suggests that the client cannot upgrade to anything else, which would make sense if the connection is WebSocket already.

**Reply by jhoogstraat** [reply]
The browser network tool shows xhr requests when receiving or emitting messages, so i don't think so.. 
If I remove `async_mode='threading'` then i get `upgrades: []` and a upgrade success log.

**Reply by miguelgrinberg** [reply]
Okay. Then provide a complete example and the output of `pip freeze` so that I can try to reproduce the problem here.

**Reply by jhoogstraat** [reply]
I tried to replicate the issue on a x86 linux machine and a raspberry pi, but couldn't.
This makes it seem like it's related to the hardware I was using, namely a Coral Dev Board...

The example I tested was
```python
from flask import Flask, send_file
from flask_socketio import SocketIO

app = Flask(__name__)
socketio = SocketIO(app, logger=False, engineio_logger=True,
                    async_mode='threading')

@socketio.on('sample')
def handle_json(sample):
    print('received: ' + str(sample))
    socketio.emit("sample", {"sample": "test"})

@app.route("/")
def index():
    return send_file("index.html")

if __name__ == '__main__':
    socketio.run(app, host="0.0.0.0")
```

and 
```html
<html>
<head>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/4.4.0/socket.io.min.js"
        integrity="sha512-I9p4eOeWXGM9m5GhJYd3UDUA5Lr+Epp5e4ykWFYW9hv3jZqdR92S5p+ApMSWuMaV4E+JqILepP1G9kNer4AFGQ=="
        crossorigin="anonymous" referrerpolicy="no-referrer">
        </script>
</head>
<body>
    <script type="text/javascript" charset="utf-8">
        var socket = io();
        socket.on('sample', function (msg) {
            console.log(msg)
        });
    </script>
    <button type="button" onclick="socket.emit('sample', {sample: 'test'})">Send test</button>
</body>
</html>
```

I add the output of `pip freeze` as soon as I have it.
